wake/testing/fuzzing/fuzz_shrink.py

```diff
@@ -0,0 +1,759 @@
+
+    # from wake.development.transactions import Error
+    if type(e1) == Error and type(e2) == Error:
+        # If it was the Error(TransactionRevertedError), compare message content.
+        if e1.message != e2.message:
+            return False
+
+    tb1 = traceback.extract_tb(e1.__traceback__)
+    tb2 = traceback.extract_tb(e2.__traceback__)
+
+    frame1 = None
+    for frame1 in tb1:
+        if is_relative_to(
+            Path(frame1.filename), Path.cwd()
+        ) and not is_relative_to(
+            Path(frame1.filename), Path().cwd() / "pytypes"
+        ):
+            break
+    frame2 = None
+    for frame2 in tb2:
+        if is_relative_to(
+            Path(frame2.filename), Path.cwd()
+        ) and not is_relative_to(
+            Path(frame2.filename), Path().cwd() / "pytypes"
+        ):
+            break
+
+    if frame1 is None or frame2 is None:
+        print("frame is none!!!!!!!!!!!!!!")
+        # return False
+    if frame1 is not None and frame2 is not None and (frame1.filename != frame2.filename
+            or  frame1.lineno != frame2.lineno
+            or frame1.name != frame2.name
+            ):
+        return False
+    return True
+
+
+class StateSnapShot:
+    _python_state: FuzzTest | None
+    chain_states: List[str]
+    flow_number: int | None # Current flow number
+
+    def __init__(self):
+        self._python_state = None
+        self.chain_states = []
+        self.flow_number = None
+
+    def take_snapshot(self, python_instance: FuzzTest, new_instance, chains: Tuple[Chain, ...], overwrite: bool):
+        if not overwrite:
+            assert self._python_state is None, "Python state already exists"
+            assert self.chain_states == [], "Chain state already exists"
+        else:
+            assert self._python_state is not None, "Python state (snapshot) is missing"
+            assert self.chain_states != [], "Chain state is missing"
+            assert self.flow_number is not None, "Flow number is missing"
+            print("Overwriting state ", self.flow_number, " to ", python_instance._flow_num)
+        # assert self._python_state is None, "Python state already exists"
+
+        self._python_state = new_instance
+
+        self.flow_number = python_instance._flow_num
+        self._python_state.__dict__.update(copy.deepcopy(python_instance.__dict__))
+        self.chain_states = [chain.snapshot() for chain in chains]
+
+
+    def revert(self, python_instance: FuzzTest, chains: Tuple[Chain, ...]):
+        assert self.chain_states != [], "Chain snapshot is missing"
+        assert self._python_state is not None, "Python state snapshot is missing "
+        assert self.flow_number is not None, "Flow number is missing"
+
+        python_instance.__dict__.update(copy.deepcopy(self._python_state.__dict__))
+
+        self._python_state = None
+        for temp_chain, chain in zip(self.chain_states, chains):
+            chain.revert(temp_chain)
+        self.chain_states = []
```

Comment:
As I understand it, we always do revert just to create another snapshot just after the revert.

In the case of chain `snapshot` I'm afraid it's necessary to call it again but I don't think we need to create `deepcopy` snapshot again. Could it be optimized? Also, I don't understand why is `deepcopy` used in this revert function.

Reply:
As far as I remember, Anvil creates a snapshot and returns the ID, but once the snapshot is used, it is removed, and reverting to that ID will fail.

But, the Python instance seems to be working with direct assignment.